File: src/store/offlineStore.test.js
import { describe, it, expect, beforeEach } from 'vitest';
import {
  getCollection,
  saveCollection,
  addItem,
  updateItem,
  deleteItem,
  seedInitialData,
  getQueue,
  clearStoreCache,
} from './offlineStore';

describe('offlineStore', () => {
  beforeEach(() => {
    localStorage.clear();
    clearStoreCache();
  });

  it('seeds sample data when localStorage is empty', () => {
    seedInitialData();
    const borrowers = getCollection('borrowers');
    const loans = getCollection('loans');

    expect(borrowers).toHaveLength(3);
    expect(loans).toHaveLength(3);
  });

  it('adds, updates, and deletes items properly while queuing operations', () => {
    const newItem = addItem('borrowers', { full_name: 'Test Borrower', monthly_income: 3000 });
    expect(newItem.id).toBeDefined();

    const collection = getCollection('borrowers');
    expect(collection).toHaveLength(1);
    expect(collection[0].full_name).toBe('Test Borrower');

    updateItem('borrowers', newItem.id, { full_name: 'Updated Name' });
    const updatedCollection = getCollection('borrowers');
    expect(updatedCollection[0].full_name).toBe('Updated Name');

    deleteItem('borrowers', newItem.id);
    expect(getCollection('borrowers')).toHaveLength(0);

    const queue = getQueue();
    expect(queue.length).toBeGreaterThanOrEqual(3);
  });
});

File: flutter_app/test/offline_store_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:microlend/store/offline_store.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  group('OfflineStore', () {
    setUp(() {
      SharedPreferences.setMockInitialValues({});
    });

    test('seeds 3 borrowers and 3 loans when empty', () async {
      final store = await OfflineStore.init();
      final borrowers = store.getCollection('borrowers');
      final loans = store.getCollection('loans');

      expect(borrowers.length, 3);
      expect(loans.length, 3);
    });

    test('performs CRUD and updates write queue', () async {
      final store = await OfflineStore.init();
      final newItem = await store.addItem('borrowers', {
        'full_name': 'Test User',
        'monthly_income': 5000.0,
      });

      expect(newItem['id'], isNotNull);
      final list = store.getCollection('borrowers');
      expect(list.length, 4);

      await store.updateItem('borrowers', newItem['id'], {'full_name': 'Updated User'});
      final updatedList = store.getCollection('borrowers');
      final updatedUser = updatedList.firstWhere((b) => b['id'] == newItem['id']);
      expect(updatedUser['full_name'], 'Updated User');

      await store.deleteItem('borrowers', newItem['id']);
      final finalCount = store.getCollection('borrowers').length;
      expect(finalCount, 3);

      final queue = store.getQueue();
      expect(queue.length, greaterThanOrEqualTo(3));

      final syncResult = await store.syncAll();
      expect(syncResult['success'], true);
      expect(store.getQueue().length, 0);
    });
  });
}
